File: go/problems/remove_invalid_parenthesis.go
package problems

import "strings"

type stack []int

func (s *stack) push(element int) {
	*s = append(*s, element)
}

func (s *stack) pop() int {
	if len(*s) == 0 {
		return -1
	}
	last := (*s)[len(*s)- 1]
	*s = (*s)[:len(*s)-1]
	return last
}

// https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses
func MinRemoveToMakeValid(s string) string {
	bracketStack := make(stack, 0)
	bytes := []byte(s)
	for index, char := range s {
		if char == '(' {
			bracketStack.push(index)
		} else if char == ')' {
			popped := bracketStack.pop()
			if popped == -1 {
				bytes[index] = ' '
			}
		}
	}
	for {
		popped := bracketStack.pop()
		if popped == -1 {
			break
		}
		bytes[popped] = ' '
	}
	return strings.Replace(string(bytes), " ", "", -1)
}

File: rust/src/problems/remove_invalid_parenthesis.rs
// https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses

pub fn min_remove_to_make_valid(s: String) -> String {
  let mut stack = Vec::new();
  let mut chars: Vec<char> = s.chars().collect();

  for i in 0..chars.len() {
    match chars[i] {
      '(' => stack.push(i),
      ')' => {
        if let Some(_) = stack.pop() {
          continue;
        } else {
          chars[i] = '\0';
        }
      },
      _ => continue,
    }
  }
  while let Some(i) = stack.pop() {
    chars[i] = '\0';
  }
  return chars.iter().collect::<String>()
}



